# 00. Import Libraries

### This script will include
#### 01. df_contract; data wrangling and consistency checks
#### 02. df_player; data wrangling and consistency checks
#### 03. Exporting cleaned data sets

In [1]:
#Import libraries

import pandas as pd
import numpy as np
import os

In [2]:
#create folder path
path = r'C:\Users\Victoria\Documents\veallen20\OneDrive\Desktop\Data Analytics Career Foundry\6.0 Advanced Analytics and Dashboard Design\Bambino 2.0'

In [3]:
#import 'contract.csv' to make df_contract
df_contract = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'contract.csv'), index_col = False)

In [66]:
#import 'player.csv' to make df_player
df_player = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'player.csv'), index_col = False)

# 01. df_contract

In [6]:
#check df_contract
df_contract.head(10)

Combined Key  Rank           Name                  CITY       STATE  \
0     Mike Trout, 2019     1     Mike Trout  Vineland, New Jersey  New Jersey   
1     Mike Trout, 2020     1     Mike Trout  Vineland, New Jersey  New Jersey   
2     Mike Trout, 2021     1     Mike Trout  Vineland, New Jersey  New Jersey   
3     Mike Trout, 2022     1     Mike Trout  Vineland, New Jersey  New Jersey   
4     Mike Trout, 2023     1     Mike Trout  Vineland, New Jersey  New Jersey   
5   Mookie Betts, 2021     2   Mookie Betts  Nashville, Tennessee   Tennessee   
6   Mookie Betts, 2022     2   Mookie Betts  Nashville, Tennessee   Tennessee   
7   Mookie Betts, 2023     2   Mookie Betts  Nashville, Tennessee   Tennessee   
8    Aaron Judge, 2023     3    Aaron Judge    Linden, New Jersey  California   
9  Manny Machado, 2023     4  Manny Machado      Hialeah, Florida     Florida   

         COUNTRY BAT THROW signed TEAM          TEAM LOCALE Position  \
0  United States   R     R         LAA  Anaheim, California      CF    
1  United States   R     R         LAA  Anaheim, California      CF    
2  United States   R     R         LAA  Anaheim, California      CF    
3  United States   R     R         LAA  Anaheim, California      CF    
4  United States   R     R         LAA  Anaheim, California      CF    
5  United States   R     R         LAD          Los Angeles      SS    
6  United States   R     R         LAD          Los Angeles      SS    
7  United States   R     R         LAD          Los Angeles      SS    
8  United States   R     R         NYY            The Bronx      CF    
9  United States   R     R         SDP            San Diego      3B    

   contract start  contract end  YEAR  contract length   total salary  
0            2019          2030  2019               12  $426,500,000   
1            2019          2030  2020               12  $426,500,000   
2            2019          2030  2021               12  $426,500,000   
3            2019          2030  2022               12  $426,500,000   
4            2019          2030  2023               12  $426,500,000   
5            2021          2032  2021               12  $365,000,000   
6            2021          2032  2022               12  $365,000,000   
7            2021          2032  2023               12  $365,000,000   
8            2023          2031  2023                9  $360,000,000   
9            2023          2033  2023               11  $350,000,000

In [9]:
df_contract.describe()

Rank  contract start  contract end         YEAR  contract length
count  1510.000000     1510.000000   1510.000000  1510.000000      1510.000000
mean    153.935762     2012.138411   2017.590728  2014.578808         6.152980
std      94.026507        6.593974      6.820576     6.835476         1.916698
min       1.000000     1997.000000   2001.000000  1997.000000         2.000000
25%      72.250000     2008.000000   2012.000000  2010.000000         5.000000
50%     151.000000     2013.000000   2019.000000  2016.000000         6.000000
75%     238.000000     2017.000000   2022.000000  2020.000000         7.000000
max     320.000000     2035.000000   2039.000000  2104.000000        16.000000

In [34]:
# Renaming columns to have consistent format

In [39]:
df_contract.rename(columns = {'Year' : 'year'}, inplace = True)

In [18]:
df_contract.rename(columns = {'CITY' : 'city'}, inplace = True)

In [19]:
df_contract.rename(columns = {'STATE' : 'state'}, inplace = True)

In [20]:
df_contract.rename(columns = {'COUNTRY' : 'country'}, inplace = True)

In [21]:
df_contract.rename(columns = {'BAT' : 'bats'}, inplace = True)

In [22]:
df_contract.rename(columns = {'THROW' : 'throws'}, inplace = True)

In [23]:
df_contract.rename(columns = {'signed TEAM' : 'team_signed'}, inplace = True)

In [24]:
df_contract.rename(columns = {'TEAM LOCALE' : 'team_locale'}, inplace = True)

In [25]:
df_contract.rename(columns = {'contract start' : 'contract_start'}, inplace = True)

In [26]:
df_contract.rename(columns = {'contract end' : 'contract_end'}, inplace = True)

In [27]:
df_contract.rename(columns = {'contract length' : 'contract_length'}, inplace = True)

In [28]:
df_contract.rename(columns = {'total salary' : 'total_salary'}, inplace = True)

In [29]:
df_contract.rename(columns = {'Rank' : 'rank'}, inplace = True)

In [30]:
df_contract.rename(columns = {'Combined Key' : 'combined_key'}, inplace = True)

In [31]:
df_contract.rename(columns = {'Name' : 'name'}, inplace = True)

In [37]:
df_contract.rename(columns = {'Position' : 'position'}, inplace = True)

In [40]:
df_contract.head()

combined_key  rank        name                  city       state  \
0  Mike Trout, 2019     1  Mike Trout  Vineland, New Jersey  New Jersey   
1  Mike Trout, 2020     1  Mike Trout  Vineland, New Jersey  New Jersey   
2  Mike Trout, 2021     1  Mike Trout  Vineland, New Jersey  New Jersey   
3  Mike Trout, 2022     1  Mike Trout  Vineland, New Jersey  New Jersey   
4  Mike Trout, 2023     1  Mike Trout  Vineland, New Jersey  New Jersey   

         country bats throws team_signed          team_locale position  \
0  United States    R      R         LAA  Anaheim, California      CF    
1  United States    R      R         LAA  Anaheim, California      CF    
2  United States    R      R         LAA  Anaheim, California      CF    
3  United States    R      R         LAA  Anaheim, California      CF    
4  United States    R      R         LAA  Anaheim, California      CF    

   contract_start  contract_end  year  contract_length   total_salary  
0            2019          2030  2019               12  $426,500,000   
1            2019          2030  2020               12  $426,500,000   
2            2019          2030  2021               12  $426,500,000   
3            2019          2030  2022               12  $426,500,000   
4            2019          2030  2023               12  $426,500,000

In [46]:
#check value counts for 'year' as the max value was suspicious
df_contract['year'].value_counts()

2023    124
2022     98
2020     96
2017     95
2018     93
2019     91
2016     89
2021     83
2015     78
2014     70
2013     62
2012     58
2009     55
2011     55
2008     51
2010     50
2007     43
2005     37
2006     36
2004     31
2003     30
2002     27
2001     22
2000     13
1999     10
2024      5
1998      5
2027      1
1997      1
2035      1
Name: year, dtype: int64

In [44]:
#Replace 2104 to 2014
df_contract['year'].replace({2104 : 2014}, inplace = True)

In [45]:
df_contract['year'].value_counts()

2023    124
2022     98
2020     96
2017     95
2018     93
2019     91
2016     89
2021     83
2015     78
2014     70
2013     62
2012     58
2009     55
2011     55
2008     51
2010     50
2007     43
2005     37
2006     36
2004     31
2003     30
2002     27
2001     22
2000     13
1999     10
2024      5
1998      5
2027      1
1997      1
2035      1
Name: year, dtype: int64

In [47]:
#check for missing values
df_contract.isnull().sum()

combined_key         0
rank                 0
name                 0
city                 0
state              500
country              0
bats                 0
throws               0
team_signed          0
team_locale          0
position             0
contract_start       0
contract_end         0
year                 0
contract_length      0
total_salary         0
dtype: int64

In [49]:
#check for mixed types
for col in df_contract.columns.tolist():
  weird = (df_contract[[col]].applymap(type) != df_contract[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_contract[weird]) > 0:
    print (col)

state


In [57]:
#Change data type of 'state' to string
df_contract['state'] = df_contract['state'].astype('str')

In [50]:
# Create a subset, df_nan, to view missing values in 'product_name' column
df_contract_nan = df_contract[df_contract['state'].isnull() == True]

In [51]:
df_contract_nan

combined_key  rank                name  \
10      Francisco Lindor, 2022     5    Francisco Lindor   
11      Francisco Lindor, 2023     5    Francisco Lindor   
12    Fernando Tatis Jr., 2021     6  Fernando Tatis Jr.   
13    Fernando Tatis Jr., 2022     6  Fernando Tatis Jr.   
14    Fernando Tatis Jr., 2023     6  Fernando Tatis Jr.   
...                        ...   ...                 ...   
1487       Hideki Matsui, 2008   320       Hideki Matsui   
1488       Hideki Matsui, 2009   320       Hideki Matsui   
1507       Ryan Dempster, 2009   320       Ryan Dempster   
1508       Ryan Dempster, 2010   320       Ryan Dempster   
1509       Ryan Dempster, 2011   320       Ryan Dempster   

                      city state             country bats throws team_signed  \
10     Caguas, Puerto Rico   NaN         Puerto Rico    S      R         NYM   
11     Caguas, Puerto Rico   NaN         Puerto Rico    S      R         NYM   
12    San Pedro de Macorís   NaN  Dominican Republic    R      R         SDP   
13    San Pedro de Macorís   NaN  Dominican Republic    R      R         SDP   
14    San Pedro de Macorís   NaN  Dominican Republic    R      R         SDP   
...                    ...   ...                 ...  ...    ...         ...   
1487        Nomi, Ishikawa   NaN               Japan    L      R         NYY   
1488        Nomi, Ishikawa   NaN               Japan    L      R         NYY   
1507               Sechelt   NaN              Canada    R      R         CHC   
1508               Sechelt   NaN              Canada    R      R         CHC   
1509               Sechelt   NaN              Canada    R      R         CHC   

           team_locale position  contract_start  contract_end  year  \
10    Flushing, Queens      SS             2022          2031  2022   
11    Flushing, Queens      SS             2022          2031  2023   
12           San Diego      RF             2021          2034  2021   
13           San Diego      RF             2021          2034  2022   
14           San Diego      RF             2021          2034  2023   
...                ...      ...             ...           ...   ...   
1487         The Bronx      DH             2006          2009  2008   
1488         The Bronx      DH             2006          2009  2009   
1507           Chicago      SP             2009          2012  2009   
1508           Chicago      SP             2009          2012  2010   
1509           Chicago      SP             2009          2012  2011   

      contract_length   total_salary  
10                 10  $341,000,000   
11                 10  $341,000,000   
12                 14  $340,000,000   
13                 14  $340,000,000   
14                 14  $340,000,000   
...               ...            ...  
1487                4   $52,000,000   
1488                4   $52,000,000   
1507                4   $52,000,000   
1508                4   $52,000,000   
1509                4   $52,000,000   

[500 rows x 16 columns]

In [58]:
# Check for full duplicates
df_contract_dups = df_contract[df_contract.duplicated()]

In [59]:
df_contract_dups

combined_key  rank           name        city state  country bats  \
884  Kenley Jansen, 2017   170  Kenley Jansen  Willemstad   nan  Curaçao    S   

    throws team_signed  team_locale position  contract_start  contract_end  \
884      R         LAD  Los Angeles      RP             2017          2021   

     year  contract_length  total_salary  
884  2017                5  $80,000,000

In [60]:
#Create new df_contract with no duplicates
df_contract_no_dups = df_contract.drop_duplicates()

### Observations of df_contract
The max value for 'YEAR' is 2104. I believe this is a typo for 2014, so I changed the value.

There are 500 missing or NaN values for 'state'. This is because these values correspond to players born outside of the US. No action needed.

 Renamed 'YEAR' to 'year'
 
Renamed 'CITY' to 'city'

Renamed 'STATE' to 'state'

Renamed 'COUNTRY' to 'country'

Renamed 'BAT' to 'bats'

Renamed 'THROW' to 'throws'

Renamed 'signed TEAM' to 'team_signed'

Renamed 'TEAM LOCALE' to 'team_locale'

Renamed 'contract start' to 'contract_start'

Renamed 'contract end' to 'contract_end'

Renamed 'contract length' to 'contract_length'

Renamed 'total salary' to 'total_salary'

Renamed 'Rank' to 'rank'

Renamed 'Combined Key' to 'combined_key'

Renamed 'Name' to 'name'

# 02. df_player

In [67]:
#check df_player
df_player.head(10)

Combined Key          NAME  Year  Age   Tm  Lg   G  WAR  \
0  A.J. Burnett, 1999  A.J. Burnett  1999   22  FLA  NL   7  0.3   
1  A.J. Burnett, 2000  A.J. Burnett  2000   23  FLA  NL  13  0.7   
2  A.J. Burnett, 2001  A.J. Burnett  2001   24  FLA  NL  27  1.7   
3  A.J. Burnett, 2002  A.J. Burnett  2002   25  FLA  NL  31  4.0   
4  A.J. Burnett, 2003  A.J. Burnett  2003   26  FLA  NL   4  0.1   
5  A.J. Burnett, 2004  A.J. Burnett  2004   27  FLA  NL  20  1.8   
6  A.J. Burnett, 2005  A.J. Burnett  2005   28  FLA  NL  32  3.4   
7  A.J. Burnett, 2006  A.J. Burnett  2006   29  TOR  AL  21  1.9   
8  A.J. Burnett, 2007  A.J. Burnett  2007   30  TOR  AL  25  2.4   
9  A.J. Burnett, 2008  A.J. Burnett  2008   31  TOR  AL  35  2.4   

           Salary   
0     $200,000.00   
1     $204,000.00   
2     $250,000.00   
3     $367,500.00   
4   $2,500,000.00   
5   $2,500,000.00   
6   $3,650,000.00   
7   $2,200,000.00   
8  $13,200,000.00   
9  $13,200,000.00

In [68]:
#Rename columns for consistency
df_player.rename(columns = {'Combined Key' : 'combined_key'}, inplace = True)

In [69]:
df_player.rename(columns = {'NAME' : 'name'}, inplace = True)

In [70]:
df_player.rename(columns = {'Year' : 'year'}, inplace = True)

In [71]:
df_player.rename(columns = {'Age' : 'age'}, inplace = True)

In [72]:
df_player.rename(columns = {'Tm' : 'team_played'}, inplace = True)

In [73]:
df_player.rename(columns = {'G' : 'games'}, inplace = True)

In [74]:
df_player.rename(columns = {'WAR' : 'war'}, inplace = True)

In [86]:
df_player.rename(columns = {' Salary ' : 'actual_salary'}, inplace = True)

In [77]:
df_player.rename(columns = {'Lg' : 'league'}, inplace = True)

In [87]:
# check column renames
df_player.head()

combined_key          name  year  age team_played league  games  war  \
0  A.J. Burnett, 1999  A.J. Burnett  1999   22         FLA     NL      7  0.3   
1  A.J. Burnett, 2000  A.J. Burnett  2000   23         FLA     NL     13  0.7   
2  A.J. Burnett, 2001  A.J. Burnett  2001   24         FLA     NL     27  1.7   
3  A.J. Burnett, 2002  A.J. Burnett  2002   25         FLA     NL     31  4.0   
4  A.J. Burnett, 2003  A.J. Burnett  2003   26         FLA     NL      4  0.1   

    actual_salary  
0    $200,000.00   
1    $204,000.00   
2    $250,000.00   
3    $367,500.00   
4  $2,500,000.00

In [88]:
df_player.describe()

year          age        games          war
count  3712.000000  3712.000000  3712.000000  3710.000000
mean   2011.405981    28.721983    86.799569     2.652291
std       7.766242     4.790437    56.051577     2.441600
min    1986.000000    18.000000     1.000000    -3.300000
25%    2006.000000    25.000000    32.000000     0.700000
50%    2012.000000    28.000000    89.000000     2.300000
75%    2018.000000    32.000000   145.000000     4.300000
max    2023.000000    45.000000   163.000000    11.900000

In [89]:
#check for missing values
df_player.isnull().sum()

combined_key       0
name               0
year               0
age                0
team_played        0
league             0
games              0
war                2
actual_salary    506
dtype: int64

In [90]:
#investigate missing salary values
df_player_salary_nan = df_player[df_player['actual_salary'].isnull() == True]

In [91]:
df_player_salary_nan

combined_key          name  year  age team_played league  games  \
17    A.J. Pollock, 2012  A.J. Pollock  2012   24         ARI     NL     31   
27    A.J. Pollock, 2022  A.J. Pollock  2022   34         CHW     AL    138   
28    A.J. Pollock, 2023  A.J. Pollock  2023   35         SFG     NL      5   
32     Aaron Hicks, 2015   Aaron Hicks  2015   25         MIN     AL     97   
40     Aaron Hicks, 2023   Aaron Hicks  2023   33         BAL     AL     43   
...                  ...           ...   ...  ...         ...    ...    ...   
3679    Yu Darvish, 2022    Yu Darvish  2022   35         SDP     NL     30   
3681  Zack Greinke, 2004  Zack Greinke  2004   20         KCR     AL     24   
3690  Zack Greinke, 2012  Zack Greinke  2012   28         LAA     AL     13   
3698  Zack Greinke, 2019  Zack Greinke  2019   35         HOU     AL     10   
3703  Zack Wheeler, 2013  Zack Wheeler  2013   23         NYM     NL     17   

      war actual_salary  
17   -0.1           NaN  
27    0.4           NaN  
28   -0.3           NaN  
32    1.4           NaN  
40    0.7           NaN  
...   ...           ...  
3679  4.4           NaN  
3681  3.8           NaN  
3690  1.5           NaN  
3698  1.1           NaN  
3703  1.2           NaN  

[506 rows x 9 columns]

In [92]:
#investigate missing war values
df_player_war_nan = df_player[df_player['war'].isnull() == True]

In [93]:
df_player_war_nan

combined_key          name  year  age team_played league  games  \
1826     Jim Thome, 2009     Jim Thome  2009   38         LAD     NL     17   
3448  Torii Hunter, 1997  Torii Hunter  1997   21         MIN     AL      1   

      war actual_salary  
1826  NaN           NaN  
3448  NaN           NaN

In [94]:
#check for mixed types
for col in df_player.columns.tolist():
  weird = (df_player[[col]].applymap(type) != df_player[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_player[weird]) > 0:
    print (col)

actual_salary


In [95]:
#Change data type of 'actual_salary' to string
df_player['actual_salary'] = df_player['actual_salary'].astype('str')

In [97]:
# Check for full duplicates
df_player_dups = df_player[df_player.duplicated()]

### Observations of df_player

All descriptive statistics are within normal range

Due to the amount of missing values in actual_salary, 13.6%, I'm considering removing this column. 

I'm considering removing 'name' and 'year' from this df_player before merging as they both already exist in df_contract.

After investigating the missing values in 'war' I found that for 3448, Torii Hunter, 1997, the player did not play any games. No action taken.

1826, Jim Thorne, 2009 for LAD, the player was traded during the season and I suspect this is why the war value is not available. No action taken.

There are no full duplicates.

Renamed 'Combined Key' to 'combined_key'

Renamed 'NAME' to 'name'

Renamed 'Year' to 'year'

Renamed 'Age' to 'age'

Renamed 'Tm' to 'team_played'

Renamed 'Lg' to 'league'

Renamed 'G' to 'games'

Renamed 'WAR' to 'war'

Renamed 'Salary' to 'salary'

### 03. Export wrangled and checked data

In [98]:
#Export df_player
df_player.to_csv(os.path.join(path, '02 Data', 'Cleaned Data', 'player_checked.csv'))

In [99]:
#Export df_contract_no_dups
df_contract_no_dups.to_csv(os.path.join(path, '02 Data', 'Cleaned Data', 'contract_checked.csv'))